h/api/test/uri_test.py

```diff
@@ -5,9 +9,131 @@
+    ("http://example.com/foo%5Bbar%5D", "http://example.com/foo%5Bbar%5D"),
+    ("http://example.com/foo[bar]", "http://example.com/foo%5Bbar%5D"),
+
+    # Path: ensure OTHER characters are encoded
+    ("http://example.com/كذا", "http://example.com/%D9%83%D8%B0%D8%A7"),
+    ("http://example.com/snowman/☃", "http://example.com/snowman/%E2%98%83"),
+
+    # Path: normalise case of encodings
+    ("http://example.com/case%2fnormalised", "http://example.com/case%2Fnormalised"),
+
+    # Query: remove empty
+    ("http://example.com?", "http://example.com"),
+
+    # Query: ensure UNRESERVED characters are decoded
+    ("http://example.com?foo%7Ebar=baz", "http://example.com?foo~bar=baz"),
+    ("http://example.com?foo~bar=baz", "http://example.com?foo~bar=baz"),
+    ("http://example.com?foo=bar%7Ebaz", "http://example.com?foo=bar~baz"),
+    ("http://example.com?foo=bar~baz", "http://example.com?foo=bar~baz"),
+    ("http://example.com?-._~:@!$'()*,=-._~:@!$='()*,", "http://example.com?-._~:@!$'()*,=-._~:@!$='()*,"),
+    ("http://example.com?%2D%2E%5F%7E%3A%40%21%24%27%28%29%2A%2C=%2D%2E%5F%7E%3A%40%21%24%3D%27%28%29%2A%2C", "http://example.com?-._~:@!$'()*,=-._~:@!$='()*,"),
+
+    # Query: ensure RESERVED characters are encoded
+    ("http://example.com?foo bar=baz", "http://example.com?foo+bar=baz"),
+    ("http://example.com?foo+bar=baz", "http://example.com?foo+bar=baz"),
+    ("http://example.com?foo%20bar=baz", "http://example.com?foo+bar=baz"),
+    ("http://example.com?foo=bar baz", "http://example.com?foo=bar+baz"),
+    ("http://example.com?foo=bar+baz", "http://example.com?foo=bar+baz"),
+    ("http://example.com?foo=bar%20baz", "http://example.com?foo=bar+baz"),
+    ("http://example.com?foo%5Bbar%5D=baz", "http://example.com?foo%5Bbar%5D=baz"),
+    ("http://example.com?foo[bar]=baz", "http://example.com?foo%5Bbar%5D=baz"),
+    ("http://example.com?foo=%5Bbar%5Dbaz", "http://example.com?foo=%5Bbar%5Dbaz"),
+    ("http://example.com?foo=[bar]baz", "http://example.com?foo=%5Bbar%5Dbaz"),
+
+    # Query: ensure OTHER characters are encoded
+    ("http://example.com?你好世界=γειά σου κόσμος", "http://example.com?%E4%BD%A0%E5%A5%BD%E4%B8%96%E7%95%8C=%CE%B3%CE%B5%CE%B9%CE%AC+%CF%83%CE%BF%CF%85+%CE%BA%CF%8C%CF%83%CE%BC%CE%BF%CF%82"),
+    ("http://example.com?love=♥", "http://example.com?love=%E2%99%A5"),
+
+    # Query: normalise case of encodings
+    ("http://example.com?love=%e2%99%a5", "http://example.com?love=%E2%99%A5"),
+
+    # Query: lexically sort parameters by name
+    ("http://example.com?a=1&b=2", "http://example.com?a=1&b=2"),
+    ("http://example.com?b=2&a=1", "http://example.com?a=1&b=2"),
+
+    # Query: preserve relative ordering of multiple params with the same name
+    ("http://example.com?b=2&b=3&b=1&a=1", "http://example.com?a=1&b=2&b=3&b=1"),
+    ("http://example.com?b=&b=3&b=1&a=1", "http://example.com?a=1&b=&b=3&b=1"),
+
+    # Query: remove parameters known to be irrelevant for document identity
```

Comment:
Google Analytics query params are designed to allow site owners to distinguish between visits to the same page, and so are almost *by definition* not going to be used to display different content.

But even if we ignore that fact, it's not a problem if this code identifies as the same some URLs which are in fact different. The question is whether *on balance*, Hypothesis users are better served by loading annotations for pages that differ only by these query params, or not.

Lacking data suggesting that large swathes of webmasters are completely irrational and have decided to overload `utm_*` query parameters to mean something different, I think this is the right behaviour for URI normalisation.